### ISSUE TITLE
Memory usage keeps increasing on md5sum compute

### ISSUE THREAD

**Issue by shashi-banger** [awslabs/mountpoint-s3#502]
### Mountpoint for Amazon S3 version

v1.0.1

### AWS Region

us-east-1

### Describe the running environment

Running on a local PC docker container. ALso experienced OOMKilled when running as a pod on AWS EKS

### What happened?

- Docker started in privileged mode
- mount of s3 bucket was successful
- When trying to execute md5sum on a 10GB file and monitoring `docker stats` memory usage keeps increasing steadily to 2GB and above
- Even tried to chunk by chunk processing using the following python code snippet

```python
def generate_file_md5(filepath):
    bufsize = 2**20 * 6
    buf = bytearray(bufsize)
    bufview = memoryview(buf)
    md5_hash = hashlib.md5()
    with open(filepath, 'rb', buffering=0) as f:
        while True:
            nread = f.readinto(bufview)
            if not nread:
                break
            md5_hash.update(bufview[:nread])
    return md5_hash.hexdigest()
```

same behaviour with above python code execution also

### Relevant log output

```shell
CONTAINER ID   NAME        CPU %     MEM USAGE / LIMIT     MEM %     NET I/O          BLOCK I/O    PIDS
5363c739f1c3   keen_pike   20.20%    1.071GiB / 7.772GiB   13.78%    13.8GB / 108MB   0B / 270kB   16
```

**Comment by shashi-banger** [comment]
Attaching debug logs:

[mountpoint-s3-2023-09-06T04-03-12Z.log](https://github.com/awslabs/mountpoint-s3/files/12532434/mountpoint-s3-2023-09-06T04-03-12Z.log)

**Comment by shashi-banger** [comment]
Attaching debug logs for reference

[mountpoint-s3-2023-09-06T04-03-12Z.log](https://github.com/awslabs/mountpoint-s3/files/12532532/mountpoint-s3-2023-09-06T04-03-12Z.log)

**Comment by shashi-banger** [comment]
Thank you for the response. Please check if it is worth considering a command line option or some configuration to limit the maximum prefetch size. This would allow users to set the memory limits for a container more reliably.

May be #510 fix will help. Will retry once it is fixed.

**Comment by CrawX** [comment]
I second that it would be very convenient to have an option to limit the memory usage for scenarios where memory availability is limited. I understand that this will likely impact performance, but that's still better than getting OOM killed.

If I wanted to modify that behavior to preload only 128MiB for example, I'd need to modify the constants here right? https://github.com/awslabs/mountpoint-s3/blob/7dcaee0966ca20c91d86b0d8b1388bcc72a24c38/mountpoint-s3/src/prefetch.rs#L142

**Comment by jamesbornholt** [comment]
@CrawX yes, that's the constant you'd want to modify to scale back the prefetcher's aggressiveness.

We're currently looking into a more comprehensive way to limit memory usage; hope to have more to share on that soon!

**Comment by eminden** [comment]
Thanks @dannycjones, this is great news, as a workaround we are using mountpoint by patching the max_request_size value for now. I'll be waiting for your work to be completed.

**Comment by dannycjones** [comment]
I've created this issue which is where we'll share updates on the automatic prefetcher scaling: #987.

**Comment by unexge** [comment]
[Mountpoint v1.10.0](https://github.com/awslabs/mountpoint-s3/releases/tag/mountpoint-s3-1.10.0) has been released with some prefetcher improvements and might reduce memory usage. Could you please try upgrading to see if it provides any improvements for you?

**Comment by yerzhan7** [comment]
[Mountpoint for Amazon S3 v1.24.0](https://github.com/awslabs/mountpoint-s3/releases/tag/mountpoint-s3-1.24.0) is now available and allows you configure memory usage target via new CLI flag `--memory-target`. See more in https://github.com/awslabs/mountpoint-s3/blob/main/doc/CONFIGURATION.md#configuring-memory-usage. Let us know if it resolves your issue.
